hs-bindgen-libclang/src/HsBindgen/Clang/Core.hs

```diff
@@ -325,7 +327,7 @@ newtype CXCursor_ = CXCursor_ (Ptr ())
 --   AST.
 --
 -- <https://clang.llvm.org/doxygen/structCXCursor.html>
-newtype CXCursor = CXCursor (ForeignPtr CXCursor_)
+newtype CXCursor = CXCursor (ForeignPtr ())
```

Comment:
is this change necessary?

Reply:
It brings this in line with all the other definitions, this was just a bug.

Reply:
It feels that *others* are buggy. We have

```haskell
insance IsForeignPtr CXCursor where
  type UnderlyingPtr CXCursor :: Ptr CXCursor_
  
   unwrapForeignPtr :: forall a. CXCursor -> (Ptr CXCursor_ -> IO a) -> IO a  
```

but we `coerce :: Ptr () -> Ptr CXCursor`, this is ok, there is `castPtr` and the `Ptr` (and `ForeignPtr` arguments are phantom, but it feels wrong.

Reply:
I don't think so..? 

```haskell
newtype DeriveIsForeignPtr u p = DeriveIsForeignPtr p

instance (Coercible p (ForeignPtr a), Coercible u (Ptr a))
      => IsForeignPtr (DeriveIsForeignPtr u p) where
  type UnderlyingPtr (DeriveIsForeignPtr u p) = u
```

so

```haskell
newtype CXCursor = CXCursor (ForeignPtr CXCursor_)
  deriving IsForeignPtr via DeriveIsForeignPtr CXCursor_ CXCursor
```

results in

```haskell
instance IsForeignPtr CXCursor where
  type UnderlyingPtr CXCursor = CXCursor_
```

not `Ptr CXCursor_`; note that `CXCursor_` is _already_ a `Ptr ()`; `Ptr CXCursor_` would be a pointer to a pointer. 

Am I wrong..?

(I'll experiment with your `ByteArray` suggestion though, which will change this anyway.)

Reply:
I'd argue that we should have `data CXCursor_` be a tag for the struct. And use `Ptr CXCursor_` to indicate pointer to that struct. `Ptr ()` is writing all bindings with `void *`. Fine, but not precise. `CApiFFI` has less to work with as `void *` pointers can be freely cast.

Reply:
I'd be okay with that also. I'll look at the `ByteArray` thing and also reconsider this.